File: src/main/java/info/deskchan/groovy_support/GroovyPlugin.java
package info.deskchan.groovy_support;

import groovy.lang.Script;
import info.deskchan.core.*;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

public abstract class GroovyPlugin extends Script implements Plugin {
	
	private PluginProxyInterface pluginProxy = null;
	private List<Runnable> cleanupHandlers = new ArrayList<>();
	private Path pluginDirPath;
	
	@Override
	public boolean initialize(PluginProxyInterface pluginProxy) {
		this.pluginProxy = pluginProxy;
		try {
			run();
		} catch (Exception e) {
			pluginProxy.log(e);
			return false;
		}
		return true;
	}
	
	@Override
	public void unload() {
		for (Runnable runnable : cleanupHandlers) {
			runnable.run();
		}
	}

	protected String getId() {
		return pluginProxy.getId();
	}
	
	protected void sendMessage(String tag, Object data) {
		pluginProxy.sendMessage(tag, data);
	}
	
	protected void sendMessage(String tag, Object data, ResponseListener responseListener) {
		pluginProxy.sendMessage(tag, data, responseListener);
	}

	protected void sendMessage(String tag, Object data, ResponseListener responseListener, ResponseListener returnListener) {
		pluginProxy.sendMessage(tag, data, responseListener, returnListener);
	}

	protected void addMessageListener(String tag, MessageListener listener) {
		pluginProxy.addMessageListener(tag, listener);
	}

	protected void removeMessageListener(String tag, MessageListener listener) {
		pluginProxy.removeMessageListener(tag, listener);
	}

	protected int setTimer(long delay, ResponseListener listener) {
		return pluginProxy.setTimer(delay, listener);
	}

	protected int setTimer(long delay, int count, ResponseListener listener) {
		return pluginProxy.setTimer(delay, count, listener);
	}

	protected void cancelTimer(int id) {
		pluginProxy.cancelTimer(id);
	}

	protected String getString(String key){ return pluginProxy.getString(key); }

	protected void addCleanupHandler(Runnable handler) {
		cleanupHandlers.add(handler);
	}

	protected PluginProperties getProperties() {
		return pluginProxy.getProperties();
	}

	protected Path getDataDirPath() {
		return pluginProxy.getDataDirPath();
	}
	
	protected Path getPluginDirPath() {
		return pluginDirPath;
	}

	protected void setResourceBundle(String path){ pluginProxy.setResourceBundle(getPluginDirPath().resolve(path).toString()); }

	protected void setConfigField(String key, Object value){ pluginProxy.setConfigField(key,value); }

	protected Object getConfigField(String key){ return pluginProxy.getConfigField(key); }
	
	protected void log(String text) {
		pluginProxy.log(text);
	}
	
	protected void log(Throwable e) {
		pluginProxy.log(e);
	}
	
	void setPluginDirPath(Path path) {
		pluginDirPath = path;
	}
	
}

File: src/main/kotlin/info/deskchan/groovy_support/GroovyPlugin.kt
package info.deskchan.groovy_support

import groovy.lang.Script
import info.deskchan.core.*
import java.nio.file.Path
import java.util.*

abstract class GroovyPlugin : Script(), Plugin {

    private var pluginProxy: PluginProxyInterface? = null
    private val cleanupHandlers = ArrayList<Runnable>()

    var pluginDirPath: Path? = null
        get() = pluginProxy!!.pluginDirPath
    var assetsDirPath: Path? = null
        get() = pluginProxy!!.assetsDirPath
    var rootDirPath: Path? = null
        get() = pluginProxy!!.rootDirPath
    val dataDirPath: Path
        get() = pluginProxy!!.dataDirPath

    val id: String
        get() = pluginProxy!!.getId()

    val properties: PluginProperties
        get() = pluginProxy!!.getProperties()


    override fun initialize(pluginProxy: PluginProxyInterface): Boolean {
        this.pluginProxy = pluginProxy
        try {
            run()
        } catch (e: Exception) {
            pluginProxy.log(e)
            return false
        }

        return true
    }

    override fun unload() {
        for (runnable in cleanupHandlers) {
            runnable.run()
        }
    }

    fun sendMessage(tag: String, data: Any?) {
        pluginProxy!!.sendMessage(tag, data)
    }

    fun sendMessage(tag: String, data: Any?, responseListener: ResponseListener) {
        pluginProxy!!.sendMessage(tag, data, responseListener)
    }

    fun sendMessage(tag: String, data: Any?, responseListener: ResponseListener, returnListener: ResponseListener) {
        pluginProxy!!.sendMessage(tag, data, responseListener, returnListener)
    }

    fun addMessageListener(tag: String, listener: MessageListener) {
        pluginProxy!!.addMessageListener(tag, listener)
    }

    fun removeMessageListener(tag: String, listener: MessageListener) {
        pluginProxy!!.removeMessageListener(tag, listener)
    }

     fun setTimer(delay: Long, listener: ResponseListener): Int {
        return pluginProxy!!.setTimer(delay, listener)
    }

    fun setTimer(delay: Long, count: Int, listener: ResponseListener): Int {
        return pluginProxy!!.setTimer(delay, count, listener)
    }

    fun cancelTimer(id: Int) {
        pluginProxy!!.cancelTimer(id)
    }

    fun getString(key: String): String {
        return pluginProxy!!.getString(key)
    }

    fun addCleanupHandler(handler: Runnable) {
        cleanupHandlers.add(handler)
    }

    fun setResourceBundle(path: String) {
        pluginProxy!!.setResourceBundle(pluginDirPath!!.resolve(path).toString())
    }

    fun setConfigField(key: String, value: Any) {
        pluginProxy!!.setConfigField(key, value)
    }

    fun getConfigField(key: String): Any? {
        return pluginProxy!!.getConfigField(key)
    }

    fun log(text: String) {
        pluginProxy!!.log(text)
    }

    fun log(e: Throwable) {
        pluginProxy!!.log(e)
    }

    fun setAlternative(srcTag: String, dstTag: String, priority:Int) {
        pluginProxy!!.setAlternative(srcTag, dstTag, priority)
    }

    fun deleteAlternative(srcTag: String, dstTag: String) {
        pluginProxy!!.deleteAlternative(srcTag, dstTag)
    }

    fun callNextAlternative(sender: String, tag: String, currentAlternative: String, data: Any?) {
        pluginProxy!!.callNextAlternative(sender, tag, currentAlternative, data)
    }

}
